File: pip_oc/pip_oc/ViewController.m
//
//  ViewController.m
//  pip_oc
//
//  Created by 无夜之星辰 on 2021/5/26.
//

#import "ViewController.h"
#import <AVKit/AVKit.h>
#import <Masonry/Masonry.h>

@interface ViewController () <AVPictureInPictureControllerDelegate>

@property (nonatomic, strong) UIButton *pipButton;
@property (nonatomic, strong) AVPictureInPictureController *pipController;
@property (nonatomic, strong) UIView *customView;
@property (nonatomic, strong) UITextView *textView;
@property (nonatomic, strong) CADisplayLink *displayLink;

@property (nonatomic) NSURLSession *session;
@property (nonatomic) NSURLSessionDownloadTask *downloadTask;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    if ([AVPictureInPictureController isPictureInPictureSupported]) {
        @try {
            NSError *error = nil;
            [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayback error:&error];
            [[AVAudioSession sharedInstance] setActive:YES withOptions:1 error:&error];
        } @catch (NSException *exception) {
            NSLog(@"AVAudioSession error = %@", exception.description);
        }
        [self setupPip];
        [self setupUI];
        [self setupCustomView];
        
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleEnterForeground) name:UIApplicationDidBecomeActiveNotification object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleEnterBackground) name:UIApplicationDidEnterBackgroundNotification object:nil];
    } else {
        NSLog(@"nonsupport pip mode");
    }
}

- (void)handleEnterForeground {
    NSLog(@"DidBecomeActive");
}

- (void)handleEnterBackground {
    NSLog(@"DidEnterBackground");
}

#pragma mark - pip

// config pip
- (void)setupPip {
    NSURL *url = [[NSBundle mainBundle] URLForResource:@"PortraitVideo" withExtension:@"MP4"];
    AVAsset *asset = [AVAsset assetWithURL:url];
    AVPlayerItem * item = [[AVPlayerItem alloc] initWithAsset:asset];
    
    AVPlayer *player = [[AVPlayer alloc] initWithPlayerItem:item];
    player.muted = YES;
    player.allowsExternalPlayback = YES;
    
    AVPlayerLayer * layer = [AVPlayerLayer playerLayerWithPlayer:player];
    layer.frame = CGRectMake(90, 90, 200, 200);
    [self.view.layer addSublayer:layer];
    
    self.pipController = [[AVPictureInPictureController alloc] initWithPlayerLayer: layer];
    self.pipController.delegate = self;
    // use KVC hide system controls and you can try other values
    [self.pipController setValue:[NSNumber numberWithInt:1] forKey:@"controlsStyle"];
}

#pragma mark - UI

- (void)setupUI {
    self.pipButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.view addSubview:self.pipButton];
    [self.pipButton setTitle:@"start/stop pip" forState:UIControlStateNormal];
    [self.pipButton addTarget:self action:@selector(pipButtonClicked) forControlEvents:UIControlEventTouchUpInside];
    [self.pipButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.mas_equalTo(self.view);
        make.top.mas_equalTo(300);
        make.width.mas_equalTo(200);
        make.height.mas_equalTo(40);
    }];
    
    UIButton *transformButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.view addSubview:transformButton];
    [transformButton setTitle:@"change shape" forState:UIControlStateNormal];
    [transformButton addTarget:self action:@selector(transform) forControlEvents:UIControlEventTouchUpInside];
    [transformButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.mas_equalTo(self.view);
        make.top.mas_equalTo(self.pipButton.mas_bottom).offset(30);
        make.width.mas_equalTo(200);
        make.height.mas_equalTo(40);
    }];
    
    UIButton *rotateButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.view addSubview:rotateButton];
    [rotateButton setTitle:@"rotate" forState:UIControlStateNormal];
    [rotateButton addTarget:self action:@selector(rotate) forControlEvents:UIControlEventTouchUpInside];
    [rotateButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.mas_equalTo(self.view);
        make.top.mas_equalTo(transformButton.mas_bottom).offset(30);
        make.width.mas_equalTo(200);
        make.height.mas_equalTo(40);
    }];
}

// your custom view
- (void)setupCustomView {
    self.customView = [[UIView alloc] init];
    self.customView.backgroundColor = [UIColor whiteColor];
    
    self.textView = [[UITextView alloc] init];
    [self.customView addSubview:self.textView];
    self.textView.backgroundColor = [UIColor blackColor];
    self.textView.textColor = [UIColor whiteColor];
    self.textView.font = [UIFont systemFontOfSize:10];
    self.textView.text = @"your text\nyour can add any view\nthis is your view\nit can be any view\ncool?\nYes this is cool\n泰裤辣！！！";
    self.textView.userInteractionEnabled = NO;
    [self.textView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(self.customView);
    }];
}

#pragma mark - timer

// 开启DisplayLink
- (void)startDisplayLink {
    [self stopDisplayLink];
    self.displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(move)];
    self.displayLink.preferredFramesPerSecond = 30;
    NSRunLoop *currentRunloop = [NSRunLoop currentRunLoop];
    // 使用常驻线程
    [currentRunloop addPort:[NSPort port] forMode:NSDefaultRunLoopMode];
    [self.displayLink addToRunLoop:currentRunloop forMode:NSDefaultRunLoopMode];
}

// 关闭DisplayLink
- (void)stopDisplayLink {
    if (self.displayLink) {
        [self.displayLink invalidate];
        self.displayLink = nil;
    }
}

#pragma - mark 移动

- (void)move {
    self.textView.contentOffset = CGPointMake(0, self.textView.contentOffset.y+1);
    if (self.textView.contentOffset.y > self.textView.contentSize.height) {
        self.textView.contentOffset = CGPointMake(0, -self.textView.contentSize.height);
    }
}

#pragma - mark start/stop pip

- (void)pipButtonClicked {
    if (self.pipController.isPictureInPictureActive) {
        [self.pipController stopPictureInPicture];
    } else {
        [self.pipController startPictureInPicture];
    }
}

#pragma mark - 变形

- (void)transform {
    NSString *videoName = @"";
    
    static int index = 0;
    index++;
    int i = index % 3;
    if (i == 0) {
        videoName = @"LandscapeVideo";
    } else if (i == 1) {
        videoName = @"SquareVideo";
    } else {
        videoName = @"PortraitVideo";
    }
    NSURL *url = [[NSBundle mainBundle] URLForResource:videoName withExtension:@"MP4"];
    AVAsset *asset = [AVAsset assetWithURL:url];
    AVPlayerItem * item = [[AVPlayerItem alloc] initWithAsset:asset];
    [self.pipController.playerLayer.player replaceCurrentItemWithPlayerItem:item];
}

#pragma mark - 旋转

- (void)rotate {
    static CGFloat angle = 0;
    angle += 0.5;
    UIWindow *window = [UIApplication sharedApplication].windows.firstObject;
    window.transform = CGAffineTransformMakeRotation(M_PI * angle);
    
    AVPlayerItem * currentItem = self.pipController.playerLayer.player.currentItem;
    
    NSURL *url = [[NSBundle mainBundle] URLForResource:@"SquareVideo" withExtension:@"MP4"];
    AVAsset *asset = [AVAsset assetWithURL:url];
    AVPlayerItem * item = [[AVPlayerItem alloc] initWithAsset:asset];
    [self.pipController.playerLayer.player replaceCurrentItemWithPlayerItem:item];
    
    [self.pipController.playerLayer.player replaceCurrentItemWithPlayerItem:currentItem];
    
    
    // TODO: - 旋转问题
    [self.customView removeFromSuperview];
    [window addSubview:self.customView];
    [self.customView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(window);
    }];
}

#pragma mark - 画中画 delegate

- (void)pictureInPictureControllerWillStartPictureInPicture:(AVPictureInPictureController *)pictureInPictureController {
    // 注意是 first window，不是 last window 也不是 key window
    UIWindow *firstWindow = [UIApplication sharedApplication].windows.firstObject;
    // 把自定义view放到画中画上
    [firstWindow addSubview:self.customView];
    [self.customView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(firstWindow);
    }];
}

- (void)pictureInPictureControllerDidStartPictureInPicture:(AVPictureInPictureController *)pictureInPictureController {
    [self startDisplayLink];
}

- (void)pictureInPictureControllerDidStopPictureInPicture:(AVPictureInPictureController *)pictureInPictureController {
    [self stopDisplayLink];
}

- (void)pictureInPictureController:(AVPictureInPictureController *)pictureInPictureController failedToStartPictureInPictureWithError:(NSError *)error {
    NSLog(@"%@", error.description);
}

@end

File: pip_swift/pip_swift/ViewController.swift
//
//  ViewController.swift
//  pip_swift
//
//  Created by 无夜之星辰 on 2021/5/26.
//

import UIKit
import AVKit
import SnapKit

class ViewController: UIViewController, AVPictureInPictureControllerDelegate {
    
    // 播放器
    private var playerLayer: AVPlayerLayer!
    // 画中画
    var pipController: AVPictureInPictureController!
    // 你的自定义view
    var customView: UIView!
    var textView: UITextView!
    // 开启/关闭画中画按钮
    private var pipButton: UIButton!
    // timer
    private var displayerLink: CADisplayLink!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        
        print("画中画初始化前：\(UIApplication.shared.windows)")
        
        if AVPictureInPictureController.isPictureInPictureSupported() {
            do {
                try AVAudioSession.sharedInstance().setCategory(.playback, options: .mixWithOthers)
            } catch {
                print(error)
            }
            setupUI()
            setupPlayer()
            setupPip()
            setupCustomView()
            NotificationCenter.default.addObserver(self, selector: #selector(handleEnterForeground), name: UIApplication.willEnterForegroundNotification, object: nil)
            NotificationCenter.default.addObserver(self, selector: #selector(handleEnterBackground), name: UIApplication.didEnterBackgroundNotification, object: nil)
        } else {
            print("不支持画中画")
        }
        
        UIApplication.shared.beginBackgroundTask {
            UIApplication.shared.endBackgroundTask(UIBackgroundTaskIdentifier.invalid)
        }
    }
    
    // 配置UI
    private func setupUI() {
        pipButton = UIButton(type: .system)
        pipButton.setTitle("开启/关闭 画中画", for: .normal)
        pipButton.addTarget(self, action: #selector(pipButtonClicked), for: .touchUpInside)
        view.addSubview(pipButton)
        pipButton.snp.makeConstraints { (make) in
            make.center.equalToSuperview()
            make.width.equalTo(200)
            make.height.equalTo(40)
        }
        
        let transformButton = UIButton(type: .system)
        view.addSubview(transformButton)
        transformButton.setTitle("改变窗口形状", for: .normal)
        transformButton.addTarget(self, action: #selector(transform), for: .touchUpInside)
        transformButton.snp.makeConstraints { (make) in
            make.centerX.equalToSuperview()
            make.top.equalTo(pipButton.snp.bottom).offset(30)
            make.width.equalTo(200)
            make.height.equalTo(40)
        }
        
        let rotateButton = UIButton(type: .system)
        view.addSubview(rotateButton)
        rotateButton.setTitle("旋转", for: .normal)
        rotateButton.addTarget(self, action: #selector(rotate), for: .touchUpInside)
        rotateButton.snp.makeConstraints { (make) in
            make.centerX.equalToSuperview()
            make.top.equalTo(transformButton.snp.bottom).offset(30)
            make.width.equalTo(200)
            make.height.equalTo(40)
        }
    }
    
    // 配置播放器
    private func setupPlayer() {
        playerLayer = AVPlayerLayer()
        playerLayer.frame = .init(x: 90, y: 90, width: 200, height: 150)
        
        let mp4Video = Bundle.main.url(forResource: "竖向视频", withExtension: "MP4")
        let asset = AVAsset.init(url: mp4Video!)
        let playerItem = AVPlayerItem.init(asset: asset)
        
        let player = AVPlayer.init(playerItem: playerItem)
        playerLayer.player = player
        player.isMuted = true
        player.allowsExternalPlayback = true
        player.play()
        
        view.layer.addSublayer(playerLayer)
    }
    
    // 配置画中画
    private func setupPip() {
        pipController = AVPictureInPictureController.init(playerLayer: playerLayer)!
        pipController.delegate = self
        // 隐藏播放按钮、快进快退按钮
        pipController.setValue(1, forKey: "controlsStyle")
        // 进入后台自动开启画中画（必须处于播放状态）
        if #available(iOS 14.2, *) {
            pipController.canStartPictureInPictureAutomaticallyFromInline = true
        } else {
            // Fallback on earlier versions
        }
    }
    
    // 配置自定义view
    private func setupCustomView() {
        customView = UIView()
        customView.backgroundColor = .white
        
        textView = UITextView()
        textView.text = """
            文本文本开头
            这是自定义view，想放什么放什么
            这是自定义view，想放什么放什么
            这是自定义view，想放什么放什么
            这是自定义view，想放什么放什么
            这是自定义view，想放什么放什么
            文本
            文本
            文本
            文本文本结尾
            """
        textView.backgroundColor = .black
        textView.textColor = .white
        textView.isUserInteractionEnabled = false
        customView.addSubview(textView)
        textView.snp.makeConstraints { (make) in
            make.edges.equalToSuperview()
        }
    }
    
    // 开启/关闭 画中画
    @objc private func pipButtonClicked() {
        if pipController.isPictureInPictureActive {
            pipController.stopPictureInPicture()
        } else {
            pipController.startPictureInPicture()
        }
    }
    
    // 开始滚动
    private func startTimer() {
        if displayerLink != nil {
            displayerLink.invalidate()
            displayerLink = nil
        }
        displayerLink = CADisplayLink.init(target: self, selector: #selector(move))
        displayerLink.preferredFramesPerSecond = 30
        let currentRunloop = RunLoop.current
        // 常驻线程
        currentRunloop.add(Port(), forMode: .default)
        displayerLink.add(to: currentRunloop, forMode: .default)
    }
    
    // 停止滚动
    private func stopTimer() {
        if displayerLink != nil {
            displayerLink.invalidate()
            displayerLink = nil
        }
    }
    
    @objc private func move() {
        let offsetY = textView.contentOffset.y
        textView.contentOffset = .init(x: 0, y: offsetY + 1)
        if textView.contentOffset.y > textView.contentSize.height {
            textView.contentOffset = .zero
        }
    }
    
    
    // MARK: - 变形
    private var index = 0
    @objc private func transform() {
        // 窗口形状由视频形状决定
        var videoName = ""
        
        index += 1
        let i = index % 3
        if i == 0 {
            videoName = "横向视频"
        } else if i == 1 {
            videoName = "方形视频"
        } else {
            videoName = "竖向视频"
        }
        let mp4Video = Bundle.main.url(forResource: videoName, withExtension: "MP4")
        let asset = AVAsset.init(url: mp4Video!)
        let playerItem = AVPlayerItem.init(asset: asset)
        
        playerLayer.player!.replaceCurrentItem(with: playerItem)
    }
    
    
    // MARK: - 旋转
    private var angle: Double = 0
    @objc private func rotate() {
        angle = angle + 0.5
        let window = UIApplication.shared.windows.first
        window?.transform = CGAffineTransform(rotationAngle: CGFloat(Double.pi * angle))
        
        let currentItem = playerLayer.player?.currentItem
        
        let mp4Video = Bundle.main.url(forResource: "方形视频", withExtension: "MP4")
        let asset = AVAsset.init(url: mp4Video!)
        let playerItem = AVPlayerItem.init(asset: asset)
        
        playerLayer.player!.replaceCurrentItem(with: playerItem)
        playerLayer.player!.replaceCurrentItem(with: currentItem)
    }
    
    
    // MARK: - 进入前后台
    
    @objc private func handleEnterForeground() {
        print("进入前台");
    }
    
    @objc private func handleEnterBackground() {
        print("进入后台");
    }
    
    
    // MARK: - Delegate
    
    // 画中画将要弹出
    func pictureInPictureControllerWillStartPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        // 打印所有window
        print("画中画初始化后：\(UIApplication.shared.windows)")
        // 注意是 first window
        if let window = UIApplication.shared.windows.first {
            // 把自定义view加到画中画上
            window.addSubview(customView)
            // 使用自动布局
            customView.snp.makeConstraints { (make) -> Void in
                make.edges.equalToSuperview()
            }
        }
    }
    
    func pictureInPictureControllerDidStartPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        startTimer()
        // 打印所有window
        print("画中画弹出后：\(UIApplication.shared.windows)")
    }
    
    func pictureInPictureControllerDidStopPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        stopTimer()
    }
    
}
